enterprise/reports: add users and events export

atereshkin opened the pull request:
<!--
👋 Hi there! Welcome.

Please check the Contributing guidelines: https://docs.goauthentik.io/docs/developer-docs/#how-can-i-contribute
-->

## Details

<!--
Explain what this PR changes, what the rationale behind the change is, if any new requirements are introduced or any breaking changes caused by this PR.

Ideally also link an Issue for context that this PR will close using `closes #`
-->
This implements users and events csv export for the enterprise version of authentik.
A new app called "authentik.enterprise.reports" has been added to house this as well as future report generation logic. 
The following new UI elements were added: "Data Exports" page (under Events), Export button in Drectory > Users, Export button in Events > Logs. The buttons are only visible when a valid enterprise license is installed.

**WARNING**
This PR does not implement any access control on the export files. It should only be merged once secure file storage is available.

---

## Checklist

-   [x] Local tests pass (`ak test authentik/`)
-   [x] The code has been formatted (`make lint-fix`)

If an API change has been made

-   [x] The API schema has been updated (`make gen-build`)

If changes to the frontend have been made

-   [x] The code has been formatted (`make web`)

If applicable

-   [x] The documentation has been updated
-   [x] The documentation has been formatted (`make docs`)

atereshkin committed:
enterprise: add users and events export (reports app)

atereshkin committed:
enterprise/reports: replace assert with AsertionError so that the assumption check is not lost when compiling to optimised byte code

BeryJu commented on authentik/core/api/users.py:
```diff
@@ -424,7 +425,7 @@ class Meta:
 
 
-class UserViewSet(UsedByMixin, ModelViewSet):
+class UserViewSet(ExportMixin, UsedByMixin, ModelViewSet):
```
It's currently more so of a desire to be able to do that, there's currently very much code that prevents this from working

atereshkin committed:
enterprise/reports: use ConditionalInheritance with ExportMixin to make reduce coupling of enterprise with the rest of authentik

atereshkin committed:
enterprise/reports: use custom iterative File to save data export instead of accessing default_storage directly, so all the FileField.save logic can run correctly (e.g. creating directories)

atereshkin committed:
enterprise/reports: change app label to simply "authentik_reports"

rissson committed:
Merge branch 'main' into csv-export

rissson committed:
wip

Signed-off-by: Marc 'risson' Schmitt <marc.schmitt@risson.space>

rissson committed:
update for new file api

Signed-off-by: Marc 'risson' Schmitt <marc.schmitt@risson.space>

PeshekDotDev committed:
sources/oauth: save returned oauth refresh tokens and add slack provider (#18501)

* sources/oauth: save returned oauth refresh tokens

* Update authentik/sources/oauth/models.py

Co-authored-by: Jens L. <jens@goauthentik.io>
Signed-off-by: Connor Peshek <connor@connorpeshek.me>

* lint

* add tests

* fix proper id setting

* update id test

---------

Signed-off-by: Connor Peshek <connor@connorpeshek.me>
Co-authored-by: connor peshek <connorpeshek@unknown1641287c8f5d.attlocal.net>
Co-authored-by: Jens L. <jens@goauthentik.io>
Co-authored-by: connor peshek <connorpeshek@connors-MacBook-Pro.local>

konradmoesch committed:
core: custom avatar url improvements (#10525)

Co-authored-by: Dominic R <dominic@sdko.org>

rissson committed:
lint

Signed-off-by: Marc 'risson' Schmitt <marc.schmitt@risson.space>

PeshekDotDev committed:
website/integrations: add salesforce (#18516)

Co-authored-by: connor peshek <connorpeshek@connors-MacBook-Pro.local>
Co-authored-by: dewi-tik <dewi@goauthentik.io>
Co-authored-by: Dominic R <dominic@sdko.org>

dominic-r commented on website/docs/users-sources/user/user_basic_operations.md:
```diff
@@ -181,3 +181,15 @@ An Admin can globally enable or disable impersonation in the [System Settings](.
+
+You can export the users in your authentik instance to a CSV file. To request a data export:
+
+1. Go to **Directory > Users** in the Admin interface.
```
```suggestion
1. Log in to authentik as an administrator and navigate to **Directory > Users**.
```

dominic-r commented on website/docs/users-sources/user/user_basic_operations.md:
```diff
@@ -181,3 +181,15 @@ An Admin can globally enable or disable impersonation in the [System Settings](.
+1. Go to **Directory > Users** in the Admin interface.
+2. Select a user folder, specify a search query, and set ordering as needed. The data export will honor these settings.
+3. Click **Export** above the user list.
+4. Note that the export is processed in the background. Once the export is ready, you will receive a notification in the Admin interface notification area.
```
```suggestion
4. Note that the export is processed in the background. Once the export is ready, you will receive a notification in the Admin interface's notification area.
```

dominic-r commented on website/docs/sys-mgmt/data-exports.md:
```diff
@@ -0,0 +1,16 @@
+The enterprise version of authentik allows you to export data (currently, users and events) in CSV format for backup or analysis purposes.
+For detailed instructions on exporting users and events, see [Export users](../users-sources/user/user_basic_operations.md#export-users) and [Export events](events/logging-events.md#export-events) respectively.
+
+You can access past data exports from the **System Management** &gt; **Data Exports** page.
```
```suggestion
You can access past data exports from the **System Management** > **Data Exports** page.
```

dominic-r commented on website/docs/sys-mgmt/events/logging-events.md:
```diff
@@ -59,3 +59,15 @@ For more examples, refer to the list of [Event actions](./event-actions.md) and
 3. For queries that include `user`, `brand`, or `context` you need to use a compound term such as `user.username` or `brand.name`.
    :::
+
+## Export events:ak-enterprise
```
```suggestion
## Export events :ak-enterprise
```

dominic-r commented on website/docs/sys-mgmt/events/logging-events.md:
```diff
@@ -59,3 +59,15 @@ For more examples, refer to the list of [Event actions](./event-actions.md) and
+
+You can export the events to a CSV file. To request a data export:
+
+1. Go to **Events > Logs** in the Admin interface.
```
```suggestion
1. Log in to authentik as an administrator and navigate to **Events > Logs**.
```

dominic-r reviewed (approved):
LGTM for docs

dominic-r committed:
Apply suggestions from code review

Signed-off-by: Dominic R <dominic@sdko.org>

dominic-r committed:
wip

BeryJu committed:
endpoints: implement endpoint stage (#18468)

* endpoints: implement endpoint stage

Signed-off-by: Jens Langhammer <jens@goauthentik.io>

* format

Signed-off-by: Jens Langhammer <jens@goauthentik.io>

* fix mismatched label

Signed-off-by: Jens Langhammer <jens@goauthentik.io>

* fix url in mdm config

Signed-off-by: Jens Langhammer <jens@goauthentik.io>

* rephrase

Signed-off-by: Jens Langhammer <jens@goauthentik.io>

* and API & UI

Signed-off-by: Jens Langhammer <jens@goauthentik.io>

* add deprecated support and deprecate gdtc

Signed-off-by: Jens Langhammer <jens@goauthentik.io>

* add stage mode

Signed-off-by: Jens Langhammer <jens@goauthentik.io>

* fixup

Signed-off-by: Jens Langhammer <jens@goauthentik.io>

* rework stage slightly, add frontend

Signed-off-by: Jens Langhammer <jens@goauthentik.io>

* include jwks, add iat and exp

Signed-off-by: Jens Langhammer <jens@goauthentik.io>

* fix tests

Signed-off-by: Jens Langhammer <jens@goauthentik.io>

* set kid

Signed-off-by: Jens Langhammer <jens@goauthentik.io>

* include device details in event list

Signed-off-by: Jens Langhammer <jens@goauthentik.io>

* format

Signed-off-by: Jens Langhammer <jens@goauthentik.io>

* implement device summary

Signed-off-by: Jens Langhammer <jens@goauthentik.io>

* add remaining tables

Signed-off-by: Jens Langhammer <jens@goauthentik.io>

* revert sanitize

Signed-off-by: Jens Langhammer <jens@goauthentik.io>

* fix uuid format issues

Signed-off-by: Jens Langhammer <jens@goauthentik.io>

---------

Signed-off-by: Jens Langhammer <jens@goauthentik.io>

BeryJu committed:
web/flows: update default background image (#18540)

Signed-off-by: Jens Langhammer <jens@goauthentik.io>

shcherbak committed:
website/integrations: add hoop.dev (#17868)

Co-authored-by: iops <iops@syneforge.com>
Co-authored-by: Dominic R <dominic@sdko.org>

GirlBossRush committed:
website: Docusaurus 3.9.2 (#18506)

BeryJu committed:
endpoints/stage: v2, better error handling, more settings (#18545)

* add options, idle fallback

Signed-off-by: Jens Langhammer <jens@goauthentik.io>

* delete other device tokens during enroll

Signed-off-by: Jens Langhammer <jens@goauthentik.io>

* better error handling

Signed-off-by: Jens Langhammer <jens@goauthentik.io>

---------

Signed-off-by: Jens Langhammer <jens@goauthentik.io>

dominic-r committed:
website: Glossary (#16007)

* website: Glossary

fix minor issues

wip

Apply suggestion from @dominic-r

Signed-off-by: Dominic R <dominic@sdko.org>

anchor to param

wip

wip

at least the lockfile changes now

sure

a-z first as tana asked

idk why i switched in the first place

wip

wip

lock

lockfiles are hard

wip

please work

no have?

Revert "no have?"

This reverts commit 743dbc1bc2900eedcc2c93af248e6afdec3688a3.

* changed to sentence-case capitalization

---------

Co-authored-by: Tana M Berry <tana@goauthentik.io>

GirlBossRush committed:
web/i18n: Locale Context Merge Branch (#18426)

* web: Update fonts to Patternfly 5 variants.

* Fix order of heading override.

* web: Flesh out locale context.

* Fix Han pattern.

* Remove comment.

* Add additional regional codes.

* Clarify comment.

* Fix typos.

* web/i18n: Add locale-specific font overrides.

* Fix stale session in locale lifecycle.

* core, web: Fix Han language codes.

* Fix warnings about invalid BCP language code.

* Build translations.

* Add locale relative labels.

* Add locale translations for Finnish and Portuguese.

* Fix XLIFF errors.

* Clean up labels.

* Tidy regions.

* Match region comment.

* Update extracted values.

* Fix locale switch not triggering on source language.

* Split labels.

* Clean up labels.

dominic-r committed:
providers/scim: cache ServiceProviderConfig (#18047)

rissson committed:
Merge branch 'main' into csv-export

rissson committed:
Merge branch 'main' into csv-export

BeryJu commented on authentik/enterprise/reports/api/reports.py:
```diff
@@ -0,0 +1,131 @@
+        You can poll that URL until `completed=true`, at which point the `file_url` property will
+        contain a URL to download
+        """
+        from authentik.enterprise.reports.tasks import generate_export
```
should be a top-level import if possible

atereshkin replied:
Sorry, it was necessary at one point, no longer is. Fixed.

BeryJu commented on authentik/enterprise/reports/models.py:
```diff
@@ -0,0 +1,125 @@
+
+    def generate(self) -> None:
+        if self.completed:
+            raise AssertionError("Data export must only be generated once")
```
should preferably be an error that inherits from `authentik.lib.sentry.SentryIgnoredException`

BeryJu commented on authentik/enterprise/reports/models.py:
```diff
@@ -0,0 +1,125 @@
+        viewset.request = request
+        queryset = viewset.get_queryset()
+
+        for backend in api_settings.DEFAULT_FILTER_BACKENDS:
```
any reason this isn't using `viewset.filter_queryset()`?

BeryJu commented on authentik/enterprise/reports/tests.py:
```diff
@@ -0,0 +1,308 @@
+
+
+@patch_license
+class TestExportPermissions(APITestCase):
```
usually when there's more than one test case class in a file I'd split it into multiple files in tests/test_xyz.py but either one is fine

BeryJu reviewed (commented):
Couple backend things, otherwise LGTM

atereshkin committed:
Update authentik/enterprise/reports/api/reports.py

Co-authored-by: Jens L. <jens@beryju.org>
Signed-off-by: Alexander Tereshkin <96586+atereshkin@users.noreply.github.com>

atereshkin committed:
enterprise/reports: got rid of unnecessary method-level import

atereshkin committed:
enterprise/reports: celan up code duplication in data export generation (invoke viewset.filter_queryset directly instead of replicating it)

atereshkin committed:
enterprise/reports: add check for app label when switching on content types

atereshkin committed:
enterprise/reports: make hyperlink field on Notification larger so it can fit the security token in the export file URL

atereshkin committed:
enterprise/reports: add is_superuser back in users export

atereshkin committed:
enterprise/reports: split tests into multiple files

atereshkin commented on authentik/enterprise/reports/models.py:
```diff
@@ -0,0 +1,125 @@
+        from authentik.core.api.users import UserViewSet
+        from authentik.events.api.events import EventViewSet
+
+        if self.content_type.model == "user":
```
Done.

atereshkin commented on authentik/enterprise/reports/serializers.py:
```diff
@@ -0,0 +1,34 @@
+
+    class Meta(UserSerializer.Meta):
+        fields = [
+            f for f in UserSerializer.Meta.fields if f not in {"groups_obj", "is_superuser"}
```
I think I originally excluded it temporarily because it was causing an N+1. And forgot to reinstate it after learning that it's acceptable. Fixed.

BeryJu reviewed (approved):
LGTM backend wise

tanberry commented on website/docs/sys-mgmt/data-exports.md:
```diff
@@ -0,0 +1,16 @@
+authentik_enterprise: true
+---
+
+The enterprise version of authentik allows you to export data (currently, users and events) in CSV format for backup or analysis purposes.
```
can we clarify what type of data exactly, so to be clear that it is not PIP? Maybe something like:
"...allows you to export data (currently, non-PIP data about user and event objects) in CSV format..."?

dewi-tik replied:
```suggestion
authentik enterprise allows you to export Users and Events data in CSV format for backup or analysis purposes.
```

dewi-tik commented on website/docs/sys-mgmt/data-exports.md:
```diff
@@ -0,0 +1,16 @@
+
+You can access past data exports from the **System Management** > **Data Exports** page.
+This page displays a list of data exports created to date. Here you can view the query used for a specific export (by expanding the export row),
+search exports by data type and user, download completed exports and delete exports you no longer need.
```
```suggestion
You can access past data exports from the **System Management** > **Data Exports** page. On this page you can view the query used for a specific export,
search exports by data type and user, download completed exports and delete exports that you no longer need.
```

dewi-tik reviewed (commented):
Some changes for the docs.

dewi-tik committed:
Apply suggestions from code review

Signed-off-by: Dewi Roberts <dewi@goauthentik.io>

dewi-tik reviewed (approved):
LGTM docs wise

dewi-tik committed:
Fixed prettier issue

GirlBossRush commented on web/src/admin/events/DataExportListPage.ts:
```diff
@@ -0,0 +1,100 @@
+    ];
+
+    renderToolbarSelected(): TemplateResult {
+        const disabled = this.selectedElements.length < 1;
```
```suggestion
        const disabled = !this.selectedElements.length;
```

kensternberg-authentik replied:
I prefer the original. It communicates clearly: "This is a container, and if it contains nothing, disable."  Boolean coercion has its places, but this is not one of them.

GirlBossRush commented on web/src/admin/reports/ExportButton.ts:
```diff
@@ -0,0 +1,55 @@
+    @property()
+    createExport: () => void = async () => {
+        throw new Error("createExport not set");
+    };
```
```suggestion
    @property({ attribute: false })
    public createExport: (() => void) | null = null
```

GirlBossRush commented on web/src/admin/reports/ExportButton.ts:
```diff
@@ -0,0 +1,55 @@
+            const apiError = await parseAPIResponseError(error);
+            await showAPIErrorMessage(apiError);
+        }
+    }
```
```suggestion
      #clickHandler = () => {
        if (typeof this.createExport !== "function") {
            throw new TypeError("`createExport` property must be a function");
        }

        return this.createExport()
          .then(() => {
            showMessage({
                level: MessageLevel.success,
                message: msg("Data export requested successfully"),
                description: msg("You will receive a notification once the data is ready"),
            });
          })
          .catch(async (error) => {
            const apiError = await parseAPIResponseError(error);
            showAPIErrorMessage(apiError);
        });
    }
```

kensternberg-authentik commented on web/src/admin/users/UserListPage.ts:
```diff
@@ -393,6 +402,11 @@ export class UserListPage extends WithBrandConfig(
                     ${msg("New Service Account")}
                 </button>
             </ak-forms-modal>
+            <ak-reports-export-button
```
I'd recommend 

```
${when(this.hasEnterpriseLicense, () => html`<ak-reports-export-button...>`, () => nothing)}
``` 

... here. Otherwise, the layout engine is obliged to treat this as a present but empty DIV, and put the usual padding and margin around it, doubling the whitespace between the two buttons next to it.

As an alternative, if you don't want to sprinkle the `.hasEnterpriseLicense` into a lot more places, you could, in `ExportButton`:

```
    // safest display setting for a button
    cachedDisplay = "inline-block";

    // memoize what the button would be if it were visible:
    connectedCallback() {
        super.connectedCallback();
        const detectedDisplay = getComputedStyle(this).display;
        if (display) {
            this.cachedDisplay = display;
        }
    }

    // Take it out of the DOM flow if it's not enabled
    willUpdate(changed: PropertyValue<this>) {
        super.willUpdate(changed);
        this.style.display = this.hasEnterpriseLicense ? this.cachedDisplay : "none";
    }

    // Don't render anything, not even an empty HTML object, if it's not enabled
    render(): TemplateResult {
        return this.hasEnterpriseLicense
            ? html` <button @click=${this.handleExportClick} class="pf-c-button pf-m-secondary">
                  ${msg("Export")}
              </button>`
            : nothing;
    }
```

kensternberg-authentik reviewed (commented):
There's a teeny-tiny bit of ugliness with the export button.

kensternberg-authentik commented on web/src/elements/sidebar/SidebarItem.ts:
```diff
@@ -199,7 +210,18 @@ export class SidebarItem extends AKElement {
+        if (this.enterprise && !this.hasEnterpriseLicense) {
+            if (!this.can(CapabilitiesEnum.IsEnterprise)) return nothing;
+            else return this.renderEnterpriseRequired();
+        }
```
I think this one is fine.  There's a difference between "That's not how we do things here" and "I wouldn't have named it that."  This feels a little bit like the latter.  And Marc's name is more expressive of intent, at least to me.

tanberry commented on website/docs/sys-mgmt/data-exports.md:
```diff
@@ -0,0 +1,15 @@
+
+## Permissions
+
+Creating or viewing a data export requires view permission on the data type being exported in addition to the respective permission for data exports.
```
Do they need view permission on the object type (user) or on "data type"? I am not sure what "data type" means in this context...?

tanberry reviewed (changes requested):
This is a great feature, and great docs to go with it, thanks @atereshkin ! A couple of nits, please, then good-to-go.

atereshkin committed:
Update web/src/admin/events/DataExportListPage.ts

Co-authored-by: Teffen Ellis <592134+GirlBossRush@users.noreply.github.com>
Signed-off-by: Alexander Tereshkin <96586+atereshkin@users.noreply.github.com>

atereshkin committed:
Update web/src/admin/events/DataExportListPage.ts

Co-authored-by: Teffen Ellis <592134+GirlBossRush@users.noreply.github.com>
Signed-off-by: Alexander Tereshkin <96586+atereshkin@users.noreply.github.com>

atereshkin committed:
Update web/src/admin/events/EventListPage.ts

Co-authored-by: Teffen Ellis <592134+GirlBossRush@users.noreply.github.com>
Signed-off-by: Alexander Tereshkin <96586+atereshkin@users.noreply.github.com>

atereshkin committed:
Update web/src/admin/reports/ExportButton.ts

Co-authored-by: Teffen Ellis <592134+GirlBossRush@users.noreply.github.com>
Signed-off-by: Alexander Tereshkin <96586+atereshkin@users.noreply.github.com>

atereshkin committed:
Update web/src/admin/reports/ExportButton.ts

Co-authored-by: Teffen Ellis <592134+GirlBossRush@users.noreply.github.com>
Signed-off-by: Alexander Tereshkin <96586+atereshkin@users.noreply.github.com>

atereshkin committed:
Update web/src/admin/users/UserListPage.ts

Co-authored-by: Teffen Ellis <592134+GirlBossRush@users.noreply.github.com>
Signed-off-by: Alexander Tereshkin <96586+atereshkin@users.noreply.github.com>

atereshkin committed:
Update web/src/elements/notifications/NotificationDrawer.ts

Co-authored-by: Teffen Ellis <592134+GirlBossRush@users.noreply.github.com>
Signed-off-by: Alexander Tereshkin <96586+atereshkin@users.noreply.github.com>

atereshkin committed:
Update web/src/elements/sidebar/SidebarItem.css

Co-authored-by: Teffen Ellis <592134+GirlBossRush@users.noreply.github.com>
Signed-off-by: Alexander Tereshkin <96586+atereshkin@users.noreply.github.com>

atereshkin committed:
enterprise/reports: resolve code review merge errors

atereshkin committed:
enterprise/reports: remove the export button from the dom flow (by settings display:none) when there's no license

atereshkin committed:
enterprise/reports: improve docs

BeryJu committed:
Merge branch 'main' into csv-export

Signed-off-by: Jens Langhammer <jens@goauthentik.io>

# Conflicts:
#	authentik/core/api/users.py
#	schema.yml

BeryJu committed:
include notification link in email

Signed-off-by: Jens Langhammer <jens@goauthentik.io>

BeryJu committed:
format

Signed-off-by: Jens Langhammer <jens@goauthentik.io>

atereshkin committed:
enterprise/reports: remove assignment assertion in ExportButton.ts

tanberry reviewed (approved):
Looks great, a fantastic feature to have, thanks @atereshkin !

BeryJu committed:
cleanup tests after perm update

Signed-off-by: Jens Langhammer <jens@goauthentik.io>

kensternberg-authentik reviewed (approved):
I still think the UI code looks fine.  Routine, other than the "enterprise" tag.
